torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -198,6 +200,53 @@ def bn2d_inference_rule(n: Node, op_type):
     else:
         raise TypeError(f'Cannot apply {op_type} with input type { arg_type} and existing type {n.type} on {n}')
 
+def calculate_hout(h_in, op_type):
+    if h_in == Dyn:
+        return Dyn
+
+    elif isinstance(h_in, int):
+        h_out = floor((h_in + (2 * op_type.padding[0] - op_type.dilation[0] *
+                               (op_type.kernel_size[0] - 1) - 1)) / op_type.stride[0]) + 1
+        return h_out
+    else:
+        raise TypeError(f'{h_in} must be a number or Dyn')
+
+def calculate_wout(w_in, op_type):
+    if w_in == Dyn:
+        return Dyn
+
+    elif isinstance(w_in, int):
+        w_out = floor((w_in + (2 * op_type.padding[1] - op_type.dilation[1] *
+                               (op_type.kernel_size[1] - 1) - 1)) /
+                      op_type.stride[1]) + 1
+        return w_out
+    else:
+        raise TypeError(f'{w_in} in {op_type} must be a number or Dyn')
+
+@register_inference_rule(Conv2d)
+def conv2d_inference_rule(n: Node, op_type):
+    assert isinstance(n.args[0], Node)
+    n.args[0].type = apply_matching(n.args[0].type, 4)
+    arg_type = n.args[0].type
+    n.type = apply_matching(n.type, 4)
```

Comment:
what does this line accomplish ? since we re-assign to n.type below

Reply:
this is for checking if the current node can be expanded to a size 4 tensor so we can check for consistency in the following conditional. I can do this directly in the conditional though instead!